docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+  message the latest at time `t+Delta` where `Delta` is a system parameter that is known. `Delta` is normally order of magnitude higher than the real communication delay 
+  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no the corresponding reply at time
+  `t + 2*Delta`, then `p` can assume that `q` is faulty. Note that the network assumptions for the consensus reactor are different 
+  (we assume partially synchronous model there). 
+
+The "fast sync" protocol is formally specified as follows:
+
+- `Correctness`: If a correct node `p` is connected to a correct node `q` for a long enough period of time, then `p` will eventually download all requested blocks from `q`.
+- `Termination`: If a set of peers of a correct node `p` is stable (no new nodes are added to the peer set) for a long enough period of time, then protocol eventually terminates. 
+- `Fairness`: A correct node `p` sends requests for blocks to all peers from its peer set.   
+
+The `Controller` can be modelled as a function with clearly defined inputs:
+
+* `State` - current state of the node. Contains data about connected peers and its behavior, pending requests, received blocks, etc.
+* `Event` - significant events in the network.
+
+producing clear outputs:
+
+* `State` - updated state of the node,
+* `Message` - signal what message to send,
+* `TimeoutTrigger` - signal that timeout should be triggered.
+
+```go
+type Event int
+
+const (
+	EventUnknown Event = iota
+	EventStatusReport
+	EventBlockRequest
+	EventNewBlock
+	EventRemovePeer
+	EventResponseTimeout
+	EventTerminationTimeout
+)
+
+type BlockRequest struct {
+	Height int64
+	PeerID p2p.ID
+}
+
+type NewBlock struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReport struct {
+	PeerID ID
+	Height int64
+}
+
+type RemovePeer struct {
+	PeerID ID
+	Height int64
+}
+
+type ResponseTimeout struct {
+	PeerID ID
+	Height int64
+}
+
+type TerminationTimeout struct {
+	Height int64
+}
+
+type Message int
+
+const (
+	MessageUnknown Message = iota
+	MessageBlockRequest
+	MessageBlockResponse
+	MessageStatusReport
+)
+
+type BlockRequestMessage struct {
+	Height int64
+	PeerID ID
+}
+
+type BlockResponseMessage struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReportMessage struct {
+	Height int64
+}
+
+type TimeoutTrigger int
+
+const (
+	TimeoutUnknown TimeoutTrigger = iota
+	TimeoutResponseTrigger
+	TimeoutTerminationTrigger
+)
+```
+
+The Controller state machine can be in two modes (states): `FastSyncMode` when
+it is trying to catch up with the network by downloading committed blocks,
+and `ConsensusMode` in which it executes Tendermint consensus protocol. We
```

Comment:
Do we even need consensus mode in blockchain package?

Reply:
We should help other peers sync-up even during consensus mode. Idea here is to emphasise that fast-sync is a mode of operation and not one shot functionality (as it is in the current code).